Convert remaining systems to use new time deriv action

nilsdeppe opened the pull request:
## Proposed changes

Change remaining evolution systems to use new dg::TimeDerivative action

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add TimeDerivativeTerms struct to GRMHD

nilsdeppe committed:
Add TimeDerivativeTerms struct to RelativisticEuler

nilsdeppe committed:
Have RelativisticEuler use dg::ComputeTimeDerivative action

nilsdeppe committed:
Add TimeDerivative struct to GeneralizedHarmonic

nilsdeppe committed:
Use dg::ComputeTimeDerivative action in GeneralizedHarmonic

nilsdeppe committed:
Switch GRMHD to use dg::ComputeTimeDerivative

nilsdeppe committed:
Add TimeDerivativeTerms struct to M1Grey

nilsdeppe committed:
Have M1Grey use dg::ComputeTimeDerivative action

nilsdeppe committed:
Remove GeneralizedHarmonic::ComputeDuDt

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/TimeDerivative.cpp:
```diff
@@ -0,0 +1,135 @@
+    const tnsr::II<DataVector, Dim, Frame::Inertial>& inv_spatial_metric,
+    const tnsr::ii<DataVector, Dim, Frame::Inertial>&
+        extrinsic_curvature) noexcept {
+  get(*pressure_star_alpha_sqrt_det_spatial_metric) =
```
I think `pressure_star_alpha_sqrt_det_spatial_metric` -> `pressure_lapse_sqrt_det_spatial_metric`, because
- it's not clear why you're calling this p* (isn't it just the pressure?)
- it's not clear why you use the name of the greek letter here, when you've used the names of the physical quantities (= lapse) elsewhere.

This change to be made in a few places, including the struct name in the .hpp

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/TimeDerivative.cpp:
```diff
@@ -0,0 +1,135 @@
+void TimeDerivative<Dim>::apply(
+    const gsl::not_null<Scalar<DataVector>*> dt_tilde_d,
+    const gsl::not_null<Scalar<DataVector>*> dt_tilde_tau,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*> dt_tilde_s,
```
Okay, I had some confusion, and had to look at the previous PR again. I got confused because this struct is called `TimeDerivative`, with an `apply` that returns via variables `dt_blah` that ... (surprise!) are _not_ in general the time derivatives, but rather are just some of the terms that make up the time derivatives.

I think it makes sense to have (as you have done) a variable `dt_blah` at the call site (the new `evolution::dg::Actions::ComputeTimeDerivative`) that you progressively add more terms to, by passing it to various helper functions.

But here, in the context of this function, the variable is never equal to the time derivative, so I don't think naming it `dt_blah` is as appropriate.

The easiest change, I think, might be to rename `dt_tilde_d` -> `dt_tilde_d_terms` or similar (and likewise in the `TimeDerivative` for each system). It might also be helpful to rename `TimeDerivative` to something like `TimeDerivativeTerms` or `TimeDerivativeHelper`?

I know, it would've been better for me to make this complaint in the previous PR 😞. What do you think of these suggestions? I'm happy to discuss any alternate suggestions

nilsdeppe replied:
How about `dt_non_flux_terms_tilde_d`? A bit long, but it's the source and non-conservative product terms that should be added.

fmahebert replied:
I like 'non flux terms', that's a good suggestion! What would you think of permuting the words to `dt_tilde_d_non_flux_terms` (or `non_flux_terms_dt_tilde_d`)? To me, at least, that's clearer that it's some terms for the d/dt, rather than the d/dt of some terms.

nilsdeppe replied:
Hmm, I'm fine with either. I can't decide which I like better. Do you have a preference?

fmahebert replied:
I'll voice a slight preference for `dt_tilde_d_non_flux_terms`, but I'd be happy with any of these (your suggestion and my two variants). Feel free to summon Larry if you'd like his opinion before making any changes :)

nilsdeppe replied:
@kidder prefers `non_flux_terms_dt_tilde_d` because it's the non-flux terms of `dt(variable)` :shrug: I'll go with that. I think any of them are a big improvement :D

fmahebert replied:
Sounds good to me!

fmahebert commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Fluxes.cpp:
```diff
@@ -73,9 +121,9 @@ void ComputeFluxes::apply(
-  DataVector& p_star_alpha_sqrt_det_g =
-      get(get<::Tags::TempScalar<1>>(temp_tensors));
-  p_star_alpha_sqrt_det_g =
+  Scalar<DataVector>& pressure_star_alpha_sqrt_det_spatial_metric =
```
Okay, here in the magnetized case I see that p != p*, but I would still suggest renaming `alpha` -> `lapse` here for consistency with other GR variable naming.

kidder replied:
ping (`pressure_star_alpha_sqrt_det_spatial_metric` -> `pressure_star_lapse_sqrt_det_spatial_metric`)

fmahebert commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/TimeDerivative.cpp:
```diff
@@ -0,0 +1,197 @@
+      get(sqrt_det_spatial_metric) * get(lapse) * get(*pressure_star);
+
+  // lapse b_i / W = lapse (B_i / W^2 + v_i (B^m v_m)
+  // reuse allocation for spatial velocity one_form
```
"spatial velocity one_form" -> either use "_" between all the words to give the variable name, or use " " everywhere to give english 😛

fmahebert commented on tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/Fluxes.py:
```diff
@@ -0,0 +1,75 @@
+                     np.outer(magnetic_field, spatial_velocity))
+
+
+def bsq(magnetic_field, spatial_metric):
```
can `bsq` -> `b_squared` if that is true? (here and elsewhere)

fmahebert commented on src/Evolution/Systems/RadiationTransport/M1Grey/TimeDerivative.hpp:
```diff
@@ -0,0 +1,88 @@
+          Frame::Inertial, NeutrinoSpecies>::type*>... dt_tilde_s,
+
+      const gsl::not_null<typename ::Tags::Flux<
+          Tags::TildeE<Frame::Inertial, NeutrinoSpecies>, tmpl::size_t<3>,
```
For my education, what is `tmpl::size_t<3>` (is it how you pass "3" in as a "typename" instead of the usual "size_t" parameter?) and why is it needed here?

nilsdeppe replied:
It's effectively a `std::integral_constant` but for reasons I have forgotten now, brigand defines its own integral_constant and `size_t` is just the `integral_constant<size_t, NUMBER>` type alias. We need to use `tmpl::size_t` because some generic algorithms end up with `tmpl` types instead of `std` types :(

fmahebert replied:
Makes sense, thanks!

fmahebert reviewed (commented):
One more involved comment that I missed on the last PR + some little comments

nilsdeppe committed:
Rename burgers time derivative

- Renames dt_vars to non_flux_terms_dt_vars
- Renames Burgers::TimeDerivative to Burgers::TimeDerivativeTerms

nilsdeppe committed:
Rename Newtonian Euler time derivatives

- Renames dt_vars to non_flux_terms_dt_vars
- Renames NewtonianEuler::TimeDerivative to NewtonianEuler::TimeDerivativeTerms

nilsdeppe commented:
@fmahebert I've pushed fixups and a couple extra commits doing renames for Newtonian hydro and Burgers. I'll do the `TimeDerivative` -> `TimeDerivativeTerms` rename once you're happy with these changes to make the fixup/squashing easier.

fmahebert commented:
@nilsdeppe LGTM, please proceed with squashing/renaming!

nilsdeppe committed:
Rename volume time derivative type alias

Rename compute_volume_time_derivative to compute_volume_time_derivative_terms in
systems

nilsdeppe commented:
@fmahebert here's what I've done:
- Added a commit at the beginning (I don't know what GitHub will choose the order to be...) that renames `compute_volume_time_derivative` to `compute_volume_time_derivative_terms` (maybe there's no `volume_`,' don't remember). This is the type alias in the systems
- There are 2 commits changing Burgers and Newtonian Euler, renaming arguments in the `TimeDerivativeTerms` classes.
- Pushed fixups for conservative systems renaming structs to `TimeDerivativeTerms`
- I found it easiest to squash in the apply function arguments renames
- I did not rename the nonconservative systems since here it is actually the full time derivative that is computed, not the terms.

fmahebert commented:
@nilsdeppe Thanks for going through the extra effort - I am happy with the renames! Feel free to squash before Larry reviews if you'd like.

nilsdeppe commented:
@fmahebert you're welcome! Thanks for being thorough and trying to make sure the code is understandable :) I've squashed things

kidder commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Sources.cpp:
```diff
@@ -93,10 +57,112 @@ struct TildeSUp {
+struct PressureStar : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+struct EnthalyTimesDensityWSquaredPlusBSquared : db::SimpleTag {
```
Enthalpy

kidder commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Sources.cpp:
```diff
@@ -93,10 +57,112 @@ struct TildeSUp {
+    const Scalar<DataVector>& specific_enthalpy,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& extrinsic_curvature,
+    const double constraint_damping_parameter) noexcept {
+  *densitized_stress = inv_spatial_metric;
```
Move this to inside the function `densitized_stress`

kidder commented on src/Evolution/Systems/RelativisticEuler/Valencia/TimeDerivativeTerms.cpp:
```diff
@@ -0,0 +1,136 @@
+/// \cond
+namespace {
+template <size_t Dim>
+void densitized_stress(
```
this isn't used

kidder commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/TimeDerivativeTerms.hpp:
```diff
@@ -0,0 +1,153 @@
+  struct PressureStar : db::SimpleTag {
+    using type = Scalar<DataVector>;
+  };
+  struct EnthalyTimesDensityWSquaredPlusBSquared : db::SimpleTag {
```
Enthalpy

kidder commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/TimeDerivativeTerms.cpp:
```diff
@@ -0,0 +1,200 @@
+      get(sqrt_det_spatial_metric) * get(lapse) * get(*pressure_star);
+
+  // lapse b_i / W = lapse (B_i / W^2 + v_i (B^m v_m)
+  // reuse allocation for spatial velocity one form
```
do we still want to do this?  It means the `spatial_velocity_one_form` is not correct exiting the function...

nilsdeppe replied:
yea, let's not do that... I'll add another temporary tag (cheap anyway since it's all one allocation) and then we can safely reuse things. Good catch!

nilsdeppe commented:
@kidder I've pushed fixups to your comments. Thanks for the review!

kidder commented:
you can squash

nilsdeppe commented:
Okay, squashed fixups :)
